app/code/Magento/Checkout/Test/Unit/Plugin/ResetQuoteAddressesTest.php

```diff
@@ -0,0 +1,172 @@
+ * Class ResetQuoteAddressesTest
+ *
+ * Test of clearing quote addresses after all items were removed.
+ */
+class ResetQuoteAddressesTest extends TestCase
+{
+    /**
+     * @var ResetQuoteAddresses
+     */
+    private $plugin;
+
+    /**
+     * @var Quote|MockObject
+     */
+    private $quoteMock;
+
+    /**
+     * @var CartExtensionInterface|MockObject
+     */
+    private $extensionAttributesMock;
+
+    /**
+     * Set Up
+     */
+    protected function setUp()
+    {
+        $this->quoteMock = $this->createPartialMock(Quote::class, [
+            'getAllAddresses',
+            'getAllVisibleItems',
+            'removeAddress',
+            'getExtensionAttributes',
+            'isVirtual',
+        ]);
+        $this->extensionAttributesMock = $this->getMockBuilder(CartExtensionInterface::class)
+            ->setMethods(
+                [
+                    'getShippingAssignments',
+                    'setShippingAssignments'
+                ]
+            )
+            ->getMockForAbstractClass();
+
+        $this->plugin = new ResetQuoteAddresses();
+    }
+
+    /**
+     * Test removing the addresses from a non empty quote
+     */
+    public function testRemovingTheAddressesFromNonEmptyQuote()
+    {
+        $quoteVisibleItems = [1, 2];
+
+        $this->quoteMock->expects($this->any())
+            ->method('getAllVisibleItems')
+            ->will($this->returnValue($quoteVisibleItems));
+        $this->quoteMock->expects($this->never())
+            ->method('getAllAddresses')
+            ->willReturnSelf();
+
+        $this->plugin->afterRemoveItem($this->quoteMock, $this->quoteMock, 1);
+    }
+
+    /**
+     * Test clearing the addresses from an empty quote
+     *
+     * @dataProvider quoteDataProvider
+     * @param bool $isVirtualQuote
+     * @param bool $quoteHasAddresses
+     * @param $extensionAttributes
+     */
+    public function testClearingTheAddressesFromEmptyQuote(
+        bool $isVirtualQuote,
+        bool $quoteHasAddresses,
+        $extensionAttributes
+    ) {
+        $quoteVisibleItems = [];
+
+        $this->quoteMock->expects($this->any())
+            ->method('getAllVisibleItems')
+            ->will($this->returnValue($quoteVisibleItems));
+
+        if ($quoteHasAddresses) {
+            $address = $this->createPartialMock(Address::class, ['getId']);
+
+            $address->expects($this->any())
+                ->method('getId')
+                ->willReturn(1);
+
+            $addresses = [$address];
+
+            $this->quoteMock->expects($this->any())
+                ->method('getAllAddresses')
+                ->will($this->returnValue($addresses));
+
+            $this->quoteMock->expects($this->exactly(count($addresses)))
+                ->method('removeAddress')
+                ->willReturnSelf();
+        } else {
+            $this->quoteMock->expects($this->any())
+                ->method('getAllAddresses')
+                ->willReturn([]);
+        }
```

Comment:
Having this conditional logic in Unit Test means that you are trying to do too much inside the single test.

Split that into two tests - for example:
- testClearingAddressesSuccessfullyFromEmptyQuoteWithAddress
- testClearingAddressesSuccessfullyFromEmptyQuoteWithoutAddress

Reply:
Moved the logic into a separated test cases. 👌